rust/src/datafusion/logical_expr.rs

```diff
@@ -112,6 +112,22 @@ pub fn resolve_expr(expr: &Expr, schema: &Schema) -> Result<Expr> {
                         right: right.clone(),
                     }));
                 }
+                (Expr::Column(l), Expr::BinaryExpr(r)) => {
+                    let Some(field) = schema.field(&l.flat_name()) else {
+                        return Err(Error::IO{message: format!("Column {} does not exist in the dataset.", l.flat_name())})
+                    };
+                    return Ok(Expr::BinaryExpr(BinaryExpr {
+                        left: left.clone(),
+                        op: *op,
+                        right: Box::new(Expr::BinaryExpr( BinaryExpr {
+                            left: coerce_expr(&r.left, &field.data_type()).map_err(|e| e).map(Box::new)?,
+                            op: r.op,
+                            right: coerce_expr(&r.right, &field.data_type()).map_err(|e| e).map(Box::new)?
```

Comment:
Why do `coerce_expr` instead of recursing into `resolve_expr`?

Reply:
resolve_expr only covers cases where
- there are binary expressions of operators or and and
or cases where one left hand side is column and right hand side is literal or vice versa.
For rest of the cases like ```x = 1 + 1``` or ```y = x - 2``` it wouldn't resolve the expr rather just let it through.
Hence I created a separate function to coerce literals in the expression to column's type.